fire/api/ts/__init__.py

```diff
@@ -0,0 +1,295 @@
+
+"""
+Modul til tidserie-håndtering.
+
+"""
+
+import datetime as dt
+from dataclasses import dataclass
+from typing import Iterable
+
+import numpy as np
+import pandas as pd
+
+from fire.api.firedb import FireDb
+
+
+# --- TEMP
+
+# ENV_DB = 'test'
+ENV_DB = 'prod'
+DB = None
+
+
+def ny_df(datacls) -> pd.DataFrame:
+    return pd.DataFrame(columns=datacls.__annotations__).astype(datacls.__annotations__)
+
+
+def dataframe(datacls):
+    """
+    Class decorator adding classmethod to create a new
+    empty Pandas DataFrame based on dataclass annotations.
+    
+    """
+    def method(cls, rows: Iterable = None) -> pd.DataFrame:
+        typedict = cls.__annotations__
+        if rows is None:
+            return pd.DataFrame(columns=typedict).astype(typedict)
+        data = (cls(*row) for row in rows)
+        return pd.DataFrame(data=data, columns=typedict).astype(typedict)
+    datacls.new_df = classmethod(method)
+    return datacls
+
+
+@dataframe
+@dataclass
+class TidsseriePost:
+    punkt_id: str = None
+    dato: np.datetime64 = None
+    kote: float = None
+    jessen_id: str = None
+    # er_jessen_punkt: bool = None
+    landsnr: str = None
+
+
+@dataframe
+@dataclass
+class ObservationsPost:
+    registrering_fra: np.datetime64 = None
+    opstillingspunkt_id: str = None
+    sigtepunkt_id: str = None
+    koteforskel: float = None
+    observationstype_id: str = None
+
+
+def get_db():
+    global DB
+    if DB is None:
+        DB = FireDb(db=ENV_DB)
+    return DB
+
+
+def fetchall(sql):
+    return get_db().session.execute(sql).all()
+
+# --- / TEMP
+
+
+def hent_tidsserie(jessen_id: str, *, get_raw_values=False) -> pd.DataFrame:
+    """
+    Hent tidsserien TS:`jessen_id`.
+
+    Punkterne i tidsserien udgør en punktgruppe.
+    Alternativt: En punktgruppe er punkter, der er en del af samme tidsserie.
+
+    Tidsserier indikeres med et SRID med formatet `TS:[JESSEN_ID]`
+    (et lokalt kotesystem), hvor `[JESSEN_ID]` er nummeret på
+    punktgruppens Jessen-punkt.
+
+    Returns
+    -------
+        Tabel med følgende kolonner:
+
+        *   punkt-ID
+        *   tidspunkt
+        *   kote i tidsseriens lokale kotesystem (`TS:[JESSEN_ID]`)
+        *   Jessen-ID
+        *   True hvis punkt er Jessenpunktet ellers 0
+    
+    """
+    sql = f"""\
+WITH jessen_punkter AS (
+    SELECT
+        p.id AS punkt_id,
+        pi.tekst AS jessen_id
+    FROM punkt p
+    JOIN punktinfo pi ON p.id = pi.punktid
+    JOIN punktinfotype pit ON pi.infotypeid = pit.infotypeid
+    WHERE
+        pit.infotype='IDENT:jessen'
+),
+landsnumre AS (
+    SELECT
+        p.id,
+        pi.tekst AS landsnr
+    FROM punkt p
+    JOIN punktinfo pi ON p.id = pi.punktid
+    JOIN punktinfotype pit ON pi.infotypeid = pit.infotypeid
+    WHERE pit.infotype = 'IDENT:landsnr'
+)
+SELECT
+    p.id AS punkt_id,
+    k.t AS dato,
+    k.z AS kote,
```

Comment:
Jeg er i hvert fald ret sikker på at disse to `AS` ikke anvendes

Reply:
@busstoptaktik selvom AS-aliasserne ikke nødvendigvis bruges direkte i SQL-koden kan det godt give mening. Det er også en måde at omdøbe kolonnerne i sit output, se fx her:

<img width="429" alt="Screenshot 2022-01-14 at 11 55 22" src="https://user-images.githubusercontent.com/13132571/149504546-addbdda2-0173-41e4-bc84-3a75d9c200d2.png">

Reply:
Praktisk!